### PR TITLE
Add InfluxDB support for traefik metrics

### PR DESCRIPTION
**Pull request by adityacs**
Add InfluxDB support for traefik metrics

This PR adds support for pushing traefik metrics to InfluxDB.

### CODE REVIEW

**Comment by adityacs** [comment]
Hi,

Kindly requesting you guys to do a quick review before I cleanup the code and rebase.

Thanks :)

File: metrics/influx.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package metrics
+
+import (
+	"time"
+    "bytes"
```
Comment (bclermont) [inline metrics/influx.go]: bad indent, you should `go fmt`

File: metrics/influx.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    "fmt"
+	"github.com/containous/traefik/log"
+	"github.com/containous/traefik/safe"
+	"github.com/containous/traefik/types"
+	kitlog "github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/metrics/influx"
+	influxdb "github.com/influxdata/influxdb/client/v2"
+)
+
+var influxClient = influx.New(map[string]string{}, influxdb.BatchPointsConfig{}, kitlog.LoggerFunc(func(keyvals ...interface{}) error {
```
Comment (bclermont) [inline metrics/influx.go]: why kitlog? traefik use logrus

Comment (adityacs) [reply metrics/influx.go]: This is dependent on go-kit/metrics/. Even other metrics related code uses the same 
https://github.com/containous/traefik/blob/master/metrics/datadog.go#L9

**Comment by ldez** [comment]
@adityacs could you run the tests locally?

- unit tests: `make test-unit`
- integration tests: `make test-integration`
- all tests: `make test`

https://github.com/containous/traefik/blob/master/CONTRIBUTING.md#tests

**Comment by adityacs** [comment]
@ldez I have added the unit tests and the code is working fine. Could you please review and approve this?

**Comment by adityacs** [comment]
@ldez Sorry. I screwed up with my git history the PR got automatically closed. Kindly re-open this PR.

**Comment by adityacs** [comment]
@bclermont @nmengin @ldez Kindly review this PR.

**Comment by ldez** [comment]
@adityacs We will review this PR, do not worry 😉 

https://github.com/containous/traefik/blob/master/MAINTAINER.md#pr-review-process

File: metrics/influx_test.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	if err := extractAndMatchMessage(t, expected, got); err != nil {
+		t.Errorf(err.Error())
+	}
+
+}
+
+type fn func()
+
+func getMessage(t *testing.T, body fn) string {
```
Comment (mmatur) [inline metrics/influx_test.go]: I can see that you have duplicated some code from [go-udp-testing](https://github.com/stvp/go-udp-testing/blob/master/udp.go#L43). We are not able to accept your PR like that. 
You have opened an [issue](https://github.com/stvp/go-udp-testing/issues/5) on go-udp-testing. Could you try do open a PR please ?

**Comment by adityacs** [comment]
@mmatur Changes has been made.

File: glide.lock [hunk 4]
```diff
@@ -95,6 +95,8 @@ imports:
 - name: github.com/coreos/etcd
   version: c400d05d0aa73e21e431c16145e558d624098018
   subpackages:
+  - Godeps/_workspace/src/github.com/ugorji/go/codec
+  - Godeps/_workspace/src/golang.org/x/net/context
```
Comment (ldez) [inline glide.lock]: could you remove those lines?

File: metrics/influx_test.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+func TestInflux(t *testing.T) {
+	udp.SetAddr(":8089")
+	// This is needed to make sure that UDP Listener listens for data a bit longer, otherwise it will quit after a millisecond
+	udp.Timeout = 5 * time.Second
+
+	influxRegistry := RegisterInflux(&types.Influx{Address: "localhost:8089", PushInterval: "1s"})
+	defer StopInflux()
+
+	if !influxRegistry.IsEnabled() {
+		t.Errorf("InfluxRegistry should return true for IsEnabled()")
```
Comment (ldez) [inline metrics/influx_test.go]: `t.Errorf` -> `t.Fatalf`

And replace `InfluxRegistry should return true for IsEnabled()` by `Influx registry must be enabled`

File: metrics/influx_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	}
+
+	expected := []string{
+		// We are only validating counts, as it is nearly impossible to validate latency, since it varies every run
```
Comment (ldez) [inline metrics/influx_test.go]: Could you move the comment above?

File: metrics/influx_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+		influxRegistry.ReqDurationHistogram().With("service", "test", "code", strconv.Itoa(http.StatusOK)).Observe(10000)
+	})
+
+	if err := extractAndMatchMessage(t, expected, got); err != nil {
+		t.Errorf(err.Error())
+	}
+
+}
+
+func extractAndMatchMessage(t *testing.T, expected []string, msg string) error {
```
Comment (ldez) [inline metrics/influx_test.go]: Could you rename, change the signature and use `t`?
```go
func assertMessage(t *testing.T, msg string, patterns []string) {
    t.Helper()
    for _, pattern := range patterns {
        re := regexp.MustCompile(pattern)
        match := re.FindStringSubmatch(msg)
        if len(match) != 2 {
            t.Errorf("Got %q %v, want %q", msg, match, pattern)
        }
    }
}
```

Please don't take over code from others: https://github.com/go-kit/kit/blob/e2b298466b32c7cd5579a9b9b07e968fc9d9452c/metrics/influx/example_test.go#L98

File: metrics/influx.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+		reqDurationHistogram: influxClient.NewHistogram(influxMetricsLatencyName),
+		retriesCounter:       influxClient.NewCounter(influxRetriesTotalName),
+	}
+}
+
+// initInfluxTicker initializes metrics pusher and creates a influxClient if not created already
+func initInfluxTicker(config *types.Influx) *time.Ticker {
+	address := config.Address
+	if len(address) == 0 {
+		address = "localhost:8089"
```
Comment (ldez) [inline metrics/influx.go]: https://docs.influxdata.com/influxdb/v0.9/concepts/api/#ports

File: metrics/influx.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+	pushInterval, err := time.ParseDuration(config.PushInterval)
+	if err != nil {
+		log.Warnf("Unable to parse %s into pushInterval, using 10s as default value", config.PushInterval)
+		pushInterval = 10 * time.Second
+	}
+
+	report := time.NewTicker(pushInterval)
+
+	var buf bytes.Buffer
```
Comment (ldez) [inline metrics/influx.go]: Could you explain why this line is not in the go routine?

File: metrics/influx.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/go-kit/kit/metrics/influx"
+	influxdb "github.com/influxdata/influxdb/client/v2"
+)
+
+var influxClient = influx.New(map[string]string{}, influxdb.BatchPointsConfig{}, kitlog.LoggerFunc(func(keyvals ...interface{}) error {
+	log.Info(keyvals)
+	return nil
+}))
+
+type bufWriter struct {
```
Comment (ldez) [inline metrics/influx.go]: Could you explain why you created `bufWriter`?

File: metrics/influx_test.go (hunk 3 above)
Comment (ldez) [reply metrics/influx_test.go]: Related PR https://github.com/stvp/go-udp-testing/pull/6

File: metrics/influx.go (hunk 8 above)
Comment (adityacs) [reply metrics/influx.go]: http port default is 8086. We are creating a udp communication here. [https://docs.influxdata.com/influxdb/v1.2/administration/config/#udp](https://docs.influxdata.com/influxdb/v1.2/administration/config/#udp)

File: metrics/influx.go (hunk 10 above)
Comment (adityacs) [reply metrics/influx.go]: https://github.com/go-kit/kit/blob/master/metrics/influx/influx.go#L99

Comment (adityacs) [reply metrics/influx.go]: We can call it with different name to avoid naming confusion

Comment (ldez) [reply metrics/influx.go]: yes please

**Comment by adityacs** [comment]
@ldez Could you please validate the changes?

**Comment by adityacs** [comment]
@ldez @mmatur @emilevauge @bclermont Thanks a lot :smiley:

**Pull request merged**